### ISSUE TITLE
Move Vehicle Update into member function

### ISSUE THREAD

**Issue by duncanspumpkin** [OpenRCT2/OpenRCT2#9473]
The vehicle update functions are currently static functions in the Vehicle.cpp file. All of them should really be `rct_vehicle` only member functions and should therefore be moved to improve readability. To begin with move only the `vehicle_update` function into a public member function.
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/ride/Vehicle.cpp#L1923-L2030

Create a new public member function of the `rct_vehicle` struct called `Update`. Rename the existing `vehicle_update` function and correct all of the variables that are used in the function. After this has been completed you can continue on to move every update function into the `rct_vehicle` struct. All of the subsequent member functions should probably be private. Don't try to move them all at once take your time and split them up into a couple compilable commits, it will be easier to review and then you wont get lost.

`cable_lift_update`, `vehicle_update_measurements`, `vehicle_update_moving_to_end_of_station`,` vehicle_update_waiting_for_passengers`, `vehicle_update_waiting_to_depart`, `vehicle_update_crash`, `vehicle_update_dodgems_mode`, `vehicle_update_swinging`, `vehicle_update_simulator_operating`, `vehicle_update_top_spin_operating`, `vehicle_update_ferris_wheel_rotating`, `vehicle_update_space_rings_operating`, `vehicle_update_haunted_house_operating`, `vehicle_update_crooked_house_operating`, `vehicle_update_rotating`, `vehicle_update_departing` , `vehicle_update_travelling`, ` vehicle_update_travelling_cable_lift`, `vehicle_update_travelling_boat`, `vehicle_update_arriving`, `vehicle_update_unloading_passengers`, `vehicle_update_waiting_for_cable_lift`, `vehicle_update_showing_film`, `vehicle_update_doing_circus_show`, ` vehicle_update_sound`

In addition make sure all local variables are in lower camelCase and all function names are in upper CamelCase. https://github.com/OpenRCT2/OpenRCT2/wiki/Coding-Style

When commiting in git try to stick to the commit message format. https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages

**Comment by dairemcdonald** [comment]
Is it okay if I work on this?

**Comment by Gymnasiast** [comment]
@dairemcdonald Certainly! These issues are specifically created for first-time contributors like you, in order to become familiar with our code.

If you need any help, you can ask us on Gitter. The link is in the readme.md file.

**Comment by UltimaBGD** [comment]
I noticed this is still outstanding, does this work still need to be done?

**Comment by Gymnasiast** [comment]
@UltimaBGD Yes, it does. If you'd like to work on it, please do so!

**Comment by UltimaBGD** [comment]
I will take a look at it!

**Comment by Gymnasiast** [comment]
@UltimaBGD Sorry, it seems I spoke too soon. A pull request is pending: https://github.com/OpenRCT2/OpenRCT2/pull/9522

**Comment by UltimaBGD** [comment]
I see, thank you for the quick response!

**Comment by TusharSariya** [comment]
hey so i was looking at this, it looks like `rct_vehicles` is renamed to `Vehicle?` . Is that the only change?

**Comment by duncanspumpkin** [comment]
Yes. I've been making a few changes to the sprites but it shouldn't change what needs done here. As said if you are having a go at this only do one or two functions and pr the result that way i can confirm its all good

**Comment by TusharSariya** [comment]
@duncanspumpkin made a pr for the vehicle_update (realized now that I should of probably re named it to Update), are you notified directly? im new to contributing.

**Comment by duncanspumpkin** [comment]
@TusharSariya I see all notifications so don't worry about that. We have left a few comments on the PR. Thanks for just doing a nice small simple one makes it easy to review and merge.

**Comment by hdpoliveira** [comment]
I'm not sure where to stop here... There are a lot of free functions in Vehicle.cpp which look like should be class methods. Should I go ahead and convert them all?

**Comment by duncanspumpkin** [comment]
@hdpoliveira that looks great. I'd leave that pr as is so that it doesn't get unmanageable. We are trying to get a big bug fix release out right now and I'm scared further work on this atm might get into merge conflict issues. 

There is always more methods that should probably become class methods. It's a never ending refactor job.

**Comment by hdpoliveira** [comment]
@duncanspumpkin Thanks a lot for taking the time to review all the changes! I went a little overboard on this, I realize it now.
I wanted to take this to the end though. I made a list of the functions which I believe should belong to the Vehicle class:

- [ ] vehicle_update_measurements
- [ ] vehicle_update_dodgems_collision
- [ ] vehicle_get_head
- [ ] vehicle_get_tail
- [ ] vehicle_peep_easteregg_here_we_are
- [ ] vehicle_test_reset
- [ ] vehicle_update_test_finish
- [ ] vehicle_invalidate_window
- [ ] vehicle_get_total_num_peeps
- [ ] vehicle_get_vehicle_entry
- [ ] vehicle_is_used_in_pairs
- [ ] vehicle_set_map_toolbar
- [ ] vehicle_get_g_forces
- [ ] vehicle_can_depart_synchronised
- [ ] vehicle_reverse_reverser_car
- [ ] vehicle_update_handle_water_splash
- [ ] vehicle_claxon
- [ ] vehicle_update_track_motion_up_stop_check
- [ ] check_and_apply_block_section_stop_site
- [ ] update_velocity
- [ ] vehicle_open_restraints
- [ ] vehicle_close_restraints
- [ ] vehicle_crash_on_water
- [ ] vehicle_crash_on_land
- [ ] vehicle_simulate_crash
- [ ] train_ready_to_depart
- [ ] vehicle_update_track_motion_mini_golf
- [ ] loc_6DA9F9
- [ ] vehicle_update_track_motion_forwards_get_new_track
- [ ] vehicle_update_track_motion_backwards_get_new_track
- [ ] vehicle_update_motion_collision_detection

How do you believe I should proceed? Should I have one PR per function? One PR per N functions?

@janisozaur thanks for the advice on commit messages! I was trying to put everything on the first line but this was not good.
For next changes, I'll do the following:

```
Part of #9473: Create Vehicle::UpdateTravellingCableLift

Converted from vehicle_update_travelling_cable_lift
```

Is this ok?

**Comment by janisozaur** [comment]
yes, looks good.

**Comment by duncanspumpkin** [comment]
Yes that's good. Just try to keep it to a couple of functions at a time.

**Comment by duncanspumpkin** [comment]
Additional functions to be converted that are assigned to @mustaphaelghoul:
- [ ] `cable_lift_update_moving_to_end_of_station`
- [ ] `cable_lift_update_waiting_to_depart`
- [ ] `cable_lift_update_departing`
- [ ] `cable_lift_update_travelling`
- [ ] `cable_lift_update_arriving`

**Comment by hdpoliveira** [comment]
I have finished the changes on my list and I'm opening small PRs this time :)

During the work here, I found out more functions that should be converted too:

- [ ] vehicle_kill_all_passengers
- [ ] vehicle_trigger_on_ride_photo
- [ ] vehicle_update_block_brakes_open_previous_section
- [ ] apply_block_brakes
- [ ] apply_non_stop_block_brake
- [ ] vehicle_update_spinning_car
- [ ] vehicle_update_swinging_car
- [ ] vehicle_get_swing_amount
- [ ] sub_6DBF3E
- [ ] vehicle_update_scenery_door
- [ ] vehicle_update_handle_scenery_door
- [ ] vehicle_play_scenery_door_open_sound
- [ ] vehicle_play_scenery_door_close_sound
- [ ] VehicleUpdateGoKartAttemptSwitchLanes
- [ ] loc_6DB38B
- [ ] vehicle_update_track_motion_powered_ride_acceleration
- [ ] cable_lift_update_track_motion
- [ ] sub_6DF21B_loop
- [ ] sub_6DF01A_loop
- [ ] vehicle_get_move_info

I also saw that many of the functions use the vehicle as the train head, and iterate on next vehicles on the train, so it would be cool to have a Train class which we could iterate on, and code would be simplified like:

```cpp
for (auto& car: vehicle->AsTrain())
{
    // do something
}
```

**Comment by duncanspumpkin** [comment]
As for the train class yes thats something like what I want to happen eventually. I'm currently trying to get the higher level of that working which is the sprite (entity) lists so that you can do something like `for (auto& veh: allVehicles)` but its hit a few issues with items being deleted whilst halfway through a loop.

**Comment by sohiladwani** [comment]
Is there anything left to do here?

**Comment by tupaschoal** [comment]
There probably is, but @hdpoliveira has been steadily doing them, so better syncing with him before starting to make sure your work doesn't overlap.

**Comment by sohiladwani** [comment]
@hdpoliveira would you mind pointing me to some which you havent done? I would like to attempt to tackle a few.

**Comment by hdpoliveira** [comment]
@sohiladwani you can work on sub_6DBF3E

All the other functions actually depend on one another.  For documentation, these are the functions pending:
- [ ] vehicle_update_track_motion_mini_golf https://github.com/OpenRCT2/OpenRCT2/pull/11716
- [ ] vehicle_update_track_motion_forwards_get_new_track
- [ ] vehicle_update_track_motion_backwards_get_new_track
- [ ] vehicle_update_motion_collision_detection
- [ ] vehicle_trigger_on_ride_photo
- [ ] vehicle_update_block_brakes_open_previous_section
- [ ] vehicle_update_scenery_door
- [ ] vehicle_update_handle_scenery_door
- [ ] vehicle_play_scenery_door_open_sound
- [ ] vehicle_play_scenery_door_close_sound
- [ ] VehicleUpdateGoKartAttemptSwitchLanes
- [ ] loc_6DB38B
- [ ] vehicle_get_move_info
- [ ] vehicle_get_move_info_size

I have patches for the 4 first

**Comment by hdpoliveira** [comment]
One other option is to do the same for the Ride class.
I see many functions in Ride.h that look like should be functions of the class (looking for functions for which the first parameter is a Ride*):

```
src\openrct2\ride\Ride.h:
  1090: void ride_update_satisfaction(Ride* ride, uint8_t happiness);
  1091: void ride_update_popularity(Ride* ride, uint8_t pop_amount);
  1095: void ride_construct(Ride* ride);
  1097: void ride_remove_peeps(Ride* ride);
  1098: void ride_clear_blocked_tiles(Ride* ride);
  1099: Staff* ride_get_mechanic(Ride* ride);
  1100: Staff* ride_get_assigned_mechanic(Ride* ride);
  1102: int32_t ride_get_total_time(Ride* ride);
  1103: TrackColour ride_get_track_colour(Ride* ride, int32_t colourScheme);
  1104: vehicle_colour ride_get_vehicle_colour(Ride* ride, int32_t vehicleIndex);
  1106: void ride_set_vehicle_colours_to_random_preset(Ride* ride, uint8_t preset_index);
  1108: std::pair<RideMeasurement*, rct_string_id> ride_get_measurement(Ride* ride);
  1109: void ride_breakdown_add_news_item(Ride* ride);
  1110: Peep* ride_find_closest_mechanic(Ride* ride, int32_t forInspection);
  1111: int32_t ride_is_valid_for_open(Ride* ride, int32_t goingToBeOpen, bool isApplying);
  1112: int32_t ride_is_valid_for_test(Ride* ride, int32_t status, bool isApplying);
  1113: int32_t ride_initialise_construction_window(Ride* ride);
  1121: void ride_prepare_breakdown(Ride* ride, int32_t breakdownReason);
  1122: TileElement* ride_get_station_start_track_element(Ride* ride, StationIndex stationIndex);
  1124: void ride_set_status(Ride* ride, int32_t status);
  1125: void ride_set_name(Ride* ride, const char* name, uint32_t flags);
  1128: money32 ride_get_common_price(Ride* forRide);
  1131: void ride_clear_for_construction(Ride* ride);
  1137: void invalidate_test_results(Ride* ride);
  1142: void increment_turn_count_1_element(Ride* ride, uint8_t type);
  1143: void increment_turn_count_2_elements(Ride* ride, uint8_t type);
  1144: void increment_turn_count_3_elements(Ride* ride, uint8_t type);
  1145: void increment_turn_count_4_plus_elements(Ride* ride, uint8_t type);
  1146: int32_t get_turn_count_1_element(Ride* ride, uint8_t type);
  1147: int32_t get_turn_count_2_elements(Ride* ride, uint8_t type);
  1148: int32_t get_turn_count_3_elements(Ride* ride, uint8_t type);
  1149: int32_t get_turn_count_4_plus_elements(Ride* ride, uint8_t type);
  1151: uint8_t ride_get_helix_sections(Ride* ride);
  1178: bool ride_are_all_possible_entrances_and_exits_built(Ride* ride);
  1179: void ride_fix_breakdown(Ride* ride, int32_t reliabilityIncreaseFactor);
  1183: void ride_update_vehicle_colours(Ride* ride);
  1190: void sub_6CB945(Ride* ride);
  1201: Vehicle* ride_get_broken_vehicle(Ride* ride);
  1209: bool ride_has_adjacent_station(Ride* ride);
  1210: bool ride_has_station_shelter(Ride* ride);
  1224: void ride_action_modify(Ride* ride, int32_t modifyType, int32_t flags);
  1227: void ride_clear_leftover_entrances(Ride* ride);
```

**Comment by duncanspumpkin** [comment]
Yes we still need to do a ride member function refactor as well. We can make another issue for that as this one was/is to track vehicle functions

**Comment by hdpoliveira** [comment]
I think that once my PRs are merged, this issue is done \o/

**Comment by tupaschoal** [comment]
@duncanspumpkin as per @hdpoliveira 's comments it seems like this job is done, so I'm closing this issue. Please reopen if you think something is left :)

Thanks Hudson!
